### PR TITLE
feat(python): auto-determine `index`/`columns`/`values` columns in `pivot` if one is left out, deprecate passing arguments positionally

### LINKED ISSUE #12087: Make `values` in `pl.pivot()` optional
### Description

As of version 0.19.12 in Polars, you are required to explicitly define the `index`, `columns`, and `values` parameters. However, once you have defined the `index` and `columns`, the `values` columns are essentially implied, i.e., the rest of the columns are all `values` columns. It would be more convenient if you have multiple values columns. Making the `values` parameter optional could possibly be helpful (similar to the default behavior in Pandas).

### PR DESCRIPTION
**Pull request by mcrumiller**
**THIS IS A BREAKING CHANGE**

Resolves #12087

Since `index` and `columns` are both required, there's no reason why `values` can't be optional. To do so requires moving `values` to the 3rd argument, which is a breaking change unless we wanted to make all arguments keyword args, or have the user supply `None` manually. It makes sense to me that `values` should be optional, as I think the most common usage is to use all remaining columns.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
Thanks for working on this

I think this is going to break too much code, especially without a deprecation period. Not totally sure about what to suggest instead though, other than keyword-only args. Maybe they could all be optional, and if you specify 2, the third is implied?

**Comment by mcrumiller** [comment]
@MarcoGorelli no problem. I've moved the `values` parameter back to the front and made it required, but if the user passes `None`, all available columns are used.

**Comment by MarcoGorelli** [comment]
Not sure about `values=None` having to be passed explicitly

Why not

> keyword-only args. Maybe they could all be optional, and if you specify 2, the third is implied?

**Comment by mcrumiller** [comment]
@MarcoGorelli yeah that's perfectly reasonable, just implemented.

**Comment by MarcoGorelli** [comment]
you'll need test for if less than 2 are specified, and deprecate_nonkeyword_arguments

**Comment by mcrumiller** [comment]
Hi @MarcoGorelli,

Sorry for not being thorough on this, been busy. New commit adds deprecation warning and also an error test for when < 2/3 args are supplied.

File: py-polars/polars/dataframe/frame.py [hunk 1]
```diff
@@ -7065,6 +7072,24 @@ def pivot(
         index = _expand_selectors(self, index)
         columns = _expand_selectors(self, columns)
 
+        # If only two of three values/index/columns are supplied, infer the third.
+        val_err_str = "must provide at least two of `values`, `index`, and `columns`"
+        if values == [None]:
+            if (index == [None]) or (columns == [None]):
+                raise ValueError(val_err_str)
+            used = list(set(index).union(columns))
+            values = [x for x in self.columns if x not in used and x not in used]
```
Comment (MarcoGorelli) [inline py-polars/polars/dataframe/frame.py:7081]: > `x not in used and x not in used`

is this repeated?

Comment (mcrumiller) [reply py-polars/polars/dataframe/frame.py]: Thanks for picking that up.

**Comment by mcrumiller** [comment]
Hi @MarcoGorelli thanks for taking a close look at this and catching my stupid mistakes.

@alexander-beedie I made a very minor change to `_expand_selectors` where `None`s are simply ignored instead of being added as elements to the list output. This doesn't cause any problems in any tests but I wanted your input on this since that's your domain.

**Review by MarcoGorelli — APPROVED** [review]
Personally, I'm on-board with the suggestion - it's not actually breaking, as there would be a deprecation period, and the arguments would become keyword-only (so, nobody's code would silently produce something different)

If I read `df.pivot("a", "b", "c")`, I have no idea what to expect (esp. as the order of arguments differs from `pandas.DataFrame.pivot`!), so I welcome keyword-only args

And determining the third argument from the 2 specified ones seems like a good ergonomic improvement

So, I'll approve, but this is pending
- OK from at least @stinodego on the API decision
- OK from @alexander-beedie on the selectors change

File: py-polars/tests/unit/operations/test_pivot.py [hunk 2]
```diff
@@ -301,7 +356,7 @@ def test_pivot_negative_duration() -> None:
         [pl.Series(name="value", values=range(len(df1) * len(df2)))]
     )
     assert df.pivot(
-        index="delta", columns="root", values="value", aggregate_function=None
+        values="value", index="delta", columns="root", aggregate_function=None
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/operations/test_pivot.py:359]: nitpick, but for ease of review, I'd suggest going through your PR and undoing all irrelevant changes

Comment (mcrumiller) [reply py-polars/tests/unit/operations/test_pivot.py]: Done.

**Comment by MarcoGorelli** [comment]
Looks like there's not much appetite for this one, and on the pandas side I've seen complaints that keyword-only arguments here are cumbersome

I'd suggest closing for now then, it can always be reconsidered in the future

**Comment by stinodego** [comment]
I admit I haven't really looked at this yet. First I'd have to figure out what pivot does exactly :smile: but I can come back to this soon.

**Comment by mcrumiller** [comment]
> First I'd have to figure out what pivot does exactly 😄

Pivot is a fancier `unstack` operation with aggregation, if that helps. I think the documentation could be improved with better column names to align directly with the keywords to show what's happening. Here is my improved example that I think lends clarity :

```python
import polars as pl

df = pl.DataFrame({
    # The unique values of this column determine the output "index" column
    # Think of this as the "group_by" column
    "index": ["one", "one", "two", "two", "three", "three",],

    # Each unique value becomes a new output column, so our output columns
    # will be "x", "y", and "z"
    "columns1": ["x", "x", "x", "y", "y", "y"],
    "columns2": ["z", "z", "z", "z", "z", "z"],

    # The values in this column must be aggregated somehow
    "values": [1, 2, 3, 4, 5, 6],
})

# df:
# shape: (6, 4)
# ┌───────┬──────────┬──────────┬────────┐
# │ index ┆ columns1 ┆ columns2 ┆ values │
# │ ---   ┆ ---      ┆ ---      ┆ ---    │
# │ str   ┆ str      ┆ str      ┆ i64    │
# ╞═══════╪══════════╪══════════╪════════╡
# │ one   ┆ x        ┆ z        ┆ 1      │
# │ one   ┆ x        ┆ z        ┆ 2      │
# │ two   ┆ x        ┆ z        ┆ 3      │
# │ two   ┆ y        ┆ z        ┆ 4      │
# │ three ┆ y        ┆ z        ┆ 5      │
# │ three ┆ y        ┆ z        ┆ 6      │
# └───────┴──────────┴──────────┴────────┘

out = df.pivot(
    values="values",
    columns=["columns1", "columns2"],
    index=["index"],
    aggregate_function="sum",
)

# out:
# shape: (3, 4)
# ┌───────┬──────┬──────┬─────┐
# │ index ┆ x    ┆ y    ┆ z   │
# │ ---   ┆ ---  ┆ ---  ┆ --- │
# │ str   ┆ i64  ┆ i64  ┆ i64 │
# ╞═══════╪══════╪══════╪═════╡
# │ one   ┆ 3*   ┆ null ┆ 3   │   *: aggregation where index = "one" and columns1 = "x"
# │ two   ┆ 3    ┆ 4*   ┆ 7   │   *: aggregation where index = "two" and columns1 = "y"
# │ three ┆ null ┆ 11   ┆ 11* │   *: aggregation where index = "three" and columns2 = "z"
# └───────┴──────┴──────┴─────┘
```

FYI this just led me to find a bug, whereby `pivot` can great a dataframe with duplicate column names:

```python
import polars as pl

df = pl.DataFrame({
    # The unique values of this column determine the output "index" column
    # Think of this as the "group_by" column
    "index": ["one", "one", "two", "two", "three", "three",],

    # Each unique value becomes a new output column, so our output columns
    # will be "x", "y", and "z"
    "columns1": ["x", "x", "x", "y", "y", "y"],
    "columns2": ["x", "x", "x", "z", "z", "z"],

    # The values in this column must be aggregated somehow
    "values": [1, 2, 3, 4, 5, 6],
})

out = df.pivot(
    values="values",
    columns=["columns1", "columns2"],
    index=["index"],
    aggregate_function="sum",
)

print(out)
```
```
shape: (3, 5)
┌───────┬──────┬──────┬──────┬──────┐
│ index ┆ x    ┆ y    ┆ x    ┆ z    │ <-- yikes, two 'x' columns
│ ---   ┆ ---  ┆ ---  ┆ ---  ┆ ---  │
│ str   ┆ i64  ┆ i64  ┆ i64  ┆ i64  │
╞═══════╪══════╪══════╪══════╪══════╡
│ one   ┆ 3    ┆ null ┆ 3    ┆ null │
│ two   ┆ 3    ┆ 4    ┆ 3    ┆ 4    │
│ three ┆ null ┆ 11   ┆ null ┆ 11   │
└───────┴──────┴──────┴──────┴──────┘
```

**Pull request closed without merging**